### DISCUSSION TITLE
전체 조회 시 값이 존재하지않으면 예외로 봐야하는가?

### DISCUSSION THREAD

**Discussion by Choi-JJunho** [woowacourse-teams/2023-pium#76] [뜨거운 토론]
다음과 같은 테스트를 작성하면서 한가지 의문점이 생겼습니다.

반려 식물 전체 조회로직을 수행했을 때 결과물이 존재하지 않으면 비어있는 List를 반환하는것이 좋을지 아니면 예외상황으로 둬야할지 이야기를 나눠보고 싶어졌습니다.

[놀람 최소화 원칙](https://ko.wikipedia.org/wiki/%EB%86%80%EB%9E%8C_%EC%B5%9C%EC%86%8C%ED%99%94_%EC%9B%90%EC%B9%99)을 봤을 때는 사용자가 식물을 가지고 있지 않는 경우 클라이언트 입장에서 비어있는 List를 받는것이 자연스럽지 않을까? 라는 생각이 들었습니다. 

```java
@Test
void 반려_식물_전체_조회시_데이터가_없으면_예외인가() {

    DataResponse<SinglePetPlantResponse> response = petPlantService.readAll(member);

    List<SinglePetPlantResponse> data = response.getData();
    assertThat(data).isEmpty();
}
```

**Comment by Kim0914** [comment]
제 개인적인 생각으로는 정상 응답을 내려주는 것이 더 적절하지 않나 생각합니다 !

우리가 ID가 1인 회원을 찾아줘 라는 요청을 처리할 때 서버에서 리소스를 찾을 수 없는 경우에는 `404`를 내려주고는 하는데요, 
이 때는 원하는 리소스를 콕 찝어서 요청을 받기 때문에 해당 리소스를 찾지 못하는 상황은 예외 상황으로 두는 것이 적절하다고 생각합니다.

반면, 탐색 결과를 보여주는 행위는 서버에서 리소스들을 모두 탐색한 후 `일치하는 결과가 없음` 자체가 하나의 결과라고 생각합니다.
그러므로 200과 함께 빈 List를 반환하는 것이 더 적절하다고 생각합니다 !

다른 분들의 의견도 궁금합니다 ~~~~ 😙

---
추가적으로 사전 식물을 검색할 때, 해당 키워드와 일치하는 결과가 없는 경우도 
주노가 말해준 상황과 함께 고민해볼 필요가 있겠네요 !!!!

**Comment by rawfishthelgh** [comment]
저도 빈 배열을 응답하는 것에 동의합니다!
프론트엔드와 협업하는 과정에서도 예외가 아닌 빈 배열을 반환하는게 
기능적으로 처리하기 더 원활할 것이라는 생각이 드네요

**Comment by yeonkkk** [comment]
저도 빈 배열로 응답하는게 자연스러운 것 같습니다. 

먼저, 클라이언트가 전체 조회를 통해 기대하는 결과가 표면적으로는 `단순히 해당하는 데이터를 받고싶다` 일 수 있겠지만, 이는 `해당하는 데이터가 있는지 없는지도 확인하고 싶다` 라는 부분도 내포하고 있다고 생각합니다. 
그렇기 때문에 빈 배열로 응답하는 것도 요청에 대한 응답이 될 수 있다고 생각합니다.

그런데 이 부분은 프론트엔드와의 논의도 필요한 부분이 아닌가 싶습니다. 
응답을 받는 입장에서 어떤 응답을 기대하는지 확인해볼 필요가 있을 거 같아요🙂

**Comment by bassyu** [comment]
저도 빈 배열에 동의합니다!

현재 프론트에서

1. 검색 결과가 있는 경우
2. 검색 결과가 없는 경우 (빈 배열)
3. 검색 중 네트워크 오류가 난 경우

3가지로 핸들링 하기 때문에 빈 배열로 보내주시면 됩니다!

**Comment by kpeel5839** [comment]
제가 봤을 때도 빈 배열로 응답하는 것이 좋아보이네용 ^.^

**Reply by Kim0914** [reply]
고마워요 매튜 ㅎㅎ

**Comment by Choi-JJunho** [comment]
빈 배열로 응답하도록 API 구성 및 기능개발 완료되어 close하겠습니다
